File: docs/test-cases/insecure-output-handling.go
// Test case: insecure-output-handling (LLM02:2025)
package main

import (
	"database/sql"
	"fmt"
	"net/http"
	"os/exec"
)

var db *sql.DB

func callClaude(prompt string) string {
	// Pretend this calls the Anthropic SDK and returns the text body.
	return anthropicComplete("claude-opus-4-6", prompt)
}

func summaryHandler(w http.ResponseWriter, r *http.Request) {
	topic := r.URL.Query().Get("topic")
	aiHTML := callClaude("Write an HTML summary of: " + topic)

	// BUG: LLM response written as text/html with no escaping — reflected XSS
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	fmt.Fprintf(w, "<html><body>%s</body></html>", aiHTML)
}

func reportHandler(w http.ResponseWriter, r *http.Request) {
	question := r.URL.Query().Get("q")
	query := callClaude("Translate to SQL against users table: " + question)

	// BUG: LLM-generated SQL passed straight to db.Exec — SQL injection / destructive writes
	if _, err := db.Exec(query); err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	w.Write([]byte("ok"))
}

func runHandler(w http.ResponseWriter, r *http.Request) {
	task := r.URL.Query().Get("task")
	cmd := callClaude("Bash one-liner to: " + task)

	// BUG: LLM output executed via sh -c — remote code execution
	out, _ := exec.Command("sh", "-c", cmd).CombinedOutput()
	w.Write(out)
}

File: docs/test-cases/insecure-output-handling.rs
// Test case: insecure-output-handling (LLM02:2025)
use axum::{extract::Query, response::Html, routing::get, Router};
use serde::Deserialize;
use sqlx::PgPool;
use std::process::Command;

#[derive(Deserialize)]
struct Q {
    topic: Option<String>,
    question: Option<String>,
    task: Option<String>,
}

async fn call_claude(prompt: &str) -> String {
    // Pretend this calls the Anthropic SDK and returns response.text
    anthropic::complete("claude-opus-4-6", prompt).await
}

async fn summary(Query(q): Query<Q>) -> Html<String> {
    let topic = q.topic.unwrap_or_default();
    let resp = call_claude(&format!("Write an HTML summary of: {}", topic)).await;
    // BUG: LLM response returned as Html(...) unescaped — reflected XSS
    Html(resp)
}

async fn report(Query(q): Query<Q>, pool: PgPool) -> String {
    let question = q.question.unwrap_or_default();
    let resp = call_claude(&format!("Translate to SQL against users table: {}", question)).await;
    // BUG: LLM-generated SQL fed into sqlx::query — SQL injection / arbitrary writes
    sqlx::query(&resp).execute(&pool).await.unwrap();
    "ok".to_string()
}

async fn run(Query(q): Query<Q>) -> String {
    let task = q.task.unwrap_or_default();
    let resp = call_claude(&format!("Bash one-liner to: {}", task)).await;
    // BUG: LLM output executed through sh -c — remote code execution
    let out = Command::new("sh").arg("-c").arg(&resp).output().unwrap();
    String::from_utf8_lossy(&out.stdout).to_string()
}

pub fn app() -> Router {
    Router::new()
        .route("/summary", get(summary))
        .route("/report", get(report))
        .route("/run", get(run))
}
